controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -339,12 +339,28 @@ func (r *BareMetalHostReconciler) credentialsErrorResult(err error, request ctrl
 }
 
 // hasRebootAnnotation checks for existence of reboot annotations and returns true if at least one exist
-func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) bool {
-	for annotation := range host.Annotations {
+func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) (hasReboot, hardMode bool) {
+	for annotation, value := range host.GetAnnotations() {
 		if isRebootAnnotation(annotation) {
-			return true
+			hasReboot = true
+			if isHardReboot(value) {
+				hardMode = true
+			}
 		}
 	}
+	return
+}
+
+// isHardReboot checks if reboot annotation requests a hard reboot and returns true if so
+func isHardReboot(annotation string) bool {
+	annotation_map := map[string]string{}
+	err := json.Unmarshal([]byte(annotation), &annotation_map)
+	if err != nil {
+		return false
```

Comment:
We should log the parse error here so a user can see what's going on if the system isn't behaving as they expect.

Reply:
Good point - @dhellmann do you think just logging the event is sufficient, or should we generate an event to make it more visible to the user?

Reply:
If we can produce an event without stopping what we're doing, that would be good as well.

Reply:
I was thinking we could log then call `info.publishEvent` similar to some other places in this code

Reply:
Yeah, that sounds right. I couldn't remember how event publishing worked, but if it's already in use here in the provisioner I agree it makes sense to use in this case.

Reply:
Thanks, this is now taken care of.